### PR TITLE
Added example of how to change sprite at runtime

### PR DESCRIPTION
**Pull request by louiidev**
# Objective

Fixes: https://github.com/bevyengine/bevy/issues/3130

## Solution

- Added an example on how to update sprites at runtime
- Showing an example of how to alter the color value
- Showing an example of how to swap the texture

### CODE REVIEW

File: examples/2d/change_sprite_runtime.rs
Comment (alice-i-cecile) [inline examples/2d/change_sprite_runtime.rs:45]: Storing raw-ish types like this is prone to conflict in surprising ways. In real code, this should be something like a `BevyLogoLight` handle, which stores this type.

We should swap to that pattern to promote idiomatic code and more clearly show what's going on.

File: Cargo.toml
Comment (alice-i-cecile) [inline Cargo.toml:120]: I think the name "modify_sprite" is probably slightly clearer to beginners? I'll defer to you and/or Cart though.

**Comment by alice-i-cecile** [comment]
Thanks for making this! So, the other thing you can actually do is fetch the `Res<Assets<ColorMaterial>>`, and then *modify* the underlying asset using `Assets::get_mut` or `Assets::set`. This will then update all of the entities that are using that handle.

This is a subtly distinct but powerful tool (especially when working with more tweakable assets like meshes). Understanding this distinction is really what made the assets system click for me when I was first learning, so I think it's valuable to highlight :)

The other question is where do we want this example to live. I had originally envisioned two examples, one for 2D and one for 3D. After chewing on it for a bit, I think this is fundamentally an *assets* example since we're trying to explain the subtleties of the assets and handles model, and the duplication would be redundant. Do you agree with my assessment there?

**Comment by louisgjohnson** [comment]
> Thanks for making this! So, the other thing you can actually do is fetch the `Res<Assets<ColorMaterial>>`, and then _modify_ the underlying asset using `Assets::get_mut` or `Assets::set`. This will then update all of the entities that are using that handle.
> 
> This is a subtly distinct but powerful tool (especially when working with more tweakable assets like meshes). Understanding this distinction is really what made the assets system click for me when I was first learning, so I think it's valuable to highlight :)
> 
> The other question is where do we want this example to live. I had originally envisioned two examples, one for 2D and one for 3D. After chewing on it for a bit, I think this is fundamentally an _assets_ example since we're trying to explain the subtleties of the assets and handles model, and the duplication would be redundant. Do you agree with my assessment there?

Thanks for the feedback, ill work on updating this PR now, yep that definitely makes more sense 👍

**Comment by alice-i-cecile** [comment]
Pinging again on this :) It came up again today, and I was hoping we had an example ready to show a beginner.

**Comment by Nilirad** [comment]
This PR adds a new example. Adding module and item level doc comments, as described in:

- #3951 and
- #4438,

would be really useful to those who will browse examples.

File: examples/2d/change_mesh_runtime.rs
Comment (alice-i-cecile) [inline examples/2d/change_mesh_runtime.rs:53]:
```suggestion
```

I think the star will be clearer like this.

File: examples/README.md
Comment (alice-i-cecile) [inline examples/README.md:85]:
```suggestion
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
1. The 2d/change_sprite_runtime example should be removed, in favor of the one in asset.
2. The change_mesh_runtime example needs to be added to the readme and Cargo.toml
3. The PR description should be updated to include the (super cool) change_mesh_runtime example.
4. The vertex shader needs more comments.
5. The examples need a little module-style doc string description at the top.

Really cool stuff though, thanks!

If you'd like, we can split out the change_mesh_runtime example into its own PR. I won't block on this, but it'll make the review and merging process easier.

**Comment by louisgjohnson** [comment]
> 2\. runtime example needs to be added to

Hey sorry I've been away for so long, at the end of last year I started a new job and moved to the other side of the country I live in and got a little burnt out of doing open source dev, ill fix this stuff up asap :)

**Comment by louisgjohnson** [comment]
@alice-i-cecile should this still be targeted at the `pipelined-rendering` branch? seems like its gone a bit stale and a few of the examples I was using as reference aren't running

**Comment by alice-i-cecile** [comment]
@louisgjohnson another user was asking about this today. Are you able to fix this up so we can merge it? :)

**Comment by louisgjohnson** [comment]
Hey @alice-i-cecile, Just an updated:
I've removed the modify mesh example and just updated the modify sprite one.
I was having some small issues with the mesh example and to prevent the whole PR being blocked I thought I could just raise it as a separate one

**Review by alice-i-cecile — APPROVED** [review]

File: examples/asset/modify_sprite.rs
Comment (mockersf) [inline examples/asset/modify_sprite.rs:63]:
```suggestion
    let (timer, mut sprite) = query.single_mut();
```

**Comment by mockersf** [comment]
I'm not sure I like this example:
* it boils down to "if you change the value of the component, it changes what is rendered using the value of the component"
* both changes happens on the same timer on the same entity, so it's hard to see what's changing what
* as the same timer is used in two systems, it's ticked only in one which is not very clear

**Comment by louisgjohnson** [comment]
> I'm not sure I like this example:
> 
> * it boils down to "if you change the value of the component, it changes what is rendered using the value of the component"
> * both changes happens on the same timer on the same entity, so it's hard to see what's changing what
> * as the same timer is used in two systems, it's ticked only in one which is not very clear

Fair enough, I can add a seperate timer for the Color component and name them accordingly?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
On second pass:

1. I would remove the color changing completely.
2. This needs to demonstrate mutating the `Assets` collection, to change the underlying value of what the handle is pointing to.
3. This needs to clearly explain and ideally demonstrate the difference between changing what a sprite entity is pointing to (mutating the handle) versus changing the underlying data (mutating the asset).

File: examples/asset/modify_sprite.rs
Comment (Nilirad) [inline examples/asset/modify_sprite.rs:19]:
```suggestion

/// Sets up the scene, creating the sprite and loading the textures.
fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
```

File: examples/asset/modify_sprite.rs
Comment (Nilirad) [inline examples/asset/modify_sprite.rs:49]:
```suggestion

/// Changes the sprite texture when `AnimationTimer` finishes.
fn change_texture(
```

File: examples/asset/modify_sprite.rs
Comment (Nilirad) [inline examples/asset/modify_sprite.rs:62]:
```suggestion

/// Changes the sprite color when `AnimationTimer` finishes.
fn change_color(mut query: Query<(&mut AnimationTimer, &mut Sprite)>) {
```

File: examples/asset/modify_sprite.rs
Comment (Vrixyz) [inline examples/asset/modify_sprite.rs:55]: I feel like timer being ticked in this system but impacting `change_color` creates an unwanted dependancy between those systems ?

**Review by Nilirad — CHANGES_REQUESTED** [review]
I think this example needs to be improved. Here are some reasons:

- Since the dark logo is never used again after switching to the light one, it doesn't make much sense to continue storing it. The new texture should be loaded on the fly, replacing the old one. They should not be stored in a resource, possibly.
- There is not a point in changing both the texture and the color in the same example: they are two unrelated things if not for being both part of `SpriteBundle`. I think we should either have a single interactive example that showcases *all* of `SpriteBundle`'s components, or many small examples each showcasing *one* component, or group of related ones.
- I really think that the name is too generic: it should be named something like `change_sprite_texture_and_color`, which is ugly, but it's literally what it's doing.
- Visually, it's boring: there is no smooth animation (e.g. color changes abruptly) and no user interaction.

File: examples/asset/modify_sprite.rs
Comment (Vrixyz) [inline examples/asset/modify_sprite.rs:38]: Would we want to simplify ?

```suggestion
        .spawn_bundle(SpriteBundle {
            texture: first_texture,
            ..default()
        })
```

**Comment by Nilirad** [comment]
Oops, Ctrl + Enter closes the PR apparently 😅

File: examples/asset/modify_sprite.rs
Comment (louisgjohnson) [reply examples/asset/modify_sprite.rs]: Yeah good point.

File: examples/asset/modify_sprite.rs
Comment (louisgjohnson) [reply examples/asset/modify_sprite.rs:56]: Using a different timer for each system now

**Comment by NthTensor** [comment]
I don't think this example is needed, and this is quite old. I propose we close this.

**Pull request closed without merging**
